distutils/_msvccompiler.py

```diff
@@ -178,28 +178,41 @@ def _find_exe(exe, paths=None):
     return exe
 
 
-PLAT_TO_VCVARS = (
-    {
-        # Use the native MSVC host if the host platform would need expensive
-        # emulation for x86.
-        'win32': 'arm64_x86',
-        'win-amd64': 'arm64_amd64',
-        'win-arm32': 'arm64_arm',
-        'win-arm64': 'arm64',
-    }
-    if get_host_platform() == "win-arm64"
-    else {
-        # Always cross-compile from x86 to work with the lighter-weight MSVC
-        # installs that do not include native 64-bit tools.
-        'win32': 'x86',
-        'win-amd64': 'x86_amd64',
-        'win-arm32': 'x86_arm',
-        'win-arm64': 'x86_arm64',
-    }
-)
-"""
-Maps get_platform() results to values expected by vcvarsall.bat.
-"""
+_vcvars_names = {
+    'win32': 'x86',
+    'win-amd64': 'amd64',
+    'win-arm32': 'arm',
+    'win-arm64': 'arm64',
+}
+
+
+def _get_vcvars_spec(host_platform, platform):
+    """
+    Given a host platform and platform, determine the spec for vcvarsall.
+
+    Uses the native MSVC host if the host platform would need expensive
+    emulation for x86.
+
+    >>> _get_vcvars_spec('win-arm64', 'win32')
+    'arm64_x86'
+    >>> _get_vcvars_spec('win-arm64', 'win-amd64')
+    'arm64_amd64'
+
+    Always cross-compile from x86 to work with the lighter-weight MSVC
+    installs that do not include native 64-bit tools.
+
+    >>> _get_vcvars_spec('win32', 'win32')
+    'x86'
+    >>> _get_vcvars_spec('win-arm32', 'win-arm32')
+    'x86_arm'
+    >>> _get_vcvars_spec('win-amd64', 'win-arm64')
+    'x86_arm64'
+    """
+    if host_platform != 'win-arm64':
+        host_platform = 'win32'
+    vc_hp = _vcvars_names[host_platform]
+    vc_plat = _vcvars_names[platform]
+    return vc_hp if vc_hp == vc_plat else f'{vc_hp}_{vc_plat}'
```

Comment:
@jaraco looks like this specific change may have broken cross-compilation (at least in some contexts). See investigation done in https://github.com/pypa/setuptools/issues/4648#issuecomment-2351152196